File: TinyScript/JAVA/src/parser/util/ParserUtils.java
package parser.util;

import jdk.jshell.spi.ExecutionControl;
import parser.ast.ASTNode;

// 为 Parser 提供的静态的方法
public class ParserUtils {
    // 树的后序遍历，来遍历语法树表达式
    public static String toPostfixExpression(ASTNode node) throws ExecutionControl.NotImplementedException {
        // left op right -> left right op
        String leftStr = "";
        String rightStr = "";

        switch (node.getType()) {
            case BINARY_EXPR:
                leftStr = toPostfixExpression(node.getChild(0));
                rightStr = toPostfixExpression(node.getChild(1));
                return leftStr + " " + rightStr + " " + node.getLexeme().getValue();
            case VARIABLE:
            case SCALAR:
                return node.getLexeme().getValue();
        }

        throw new ExecutionControl.NotImplementedException("not impl.");
    }
}

File: TinyScript/JS/src/parser/util/ParserUtils.js
const ASTNodeTypes = require('../ast/ASTNodeTypes')

class ParserUtils {
  // 把语法生成树以后序的形式处理
  static toPostfixExpression(astNode) {
    switch (astNode.type) {
      case ASTNodeTypes.BINARY_EXPR:
        return `${ParserUtils.toPostfixExpression(astNode.getChild(0))} ${ParserUtils.toPostfixExpression(astNode.getChild(1))} ${astNode.lexeme.getValue()}`
      case ASTNodeTypes.VARIABLE:
      case ASTNodeTypes.SCALAR:
        return astNode.lexeme.getValue()
    }

    throw "not impl."
  }
}

module.exports = ParserUtils